docs/LoaderDriverInterface.md

```diff
@@ -632,15 +632,13 @@ VKAPI_ATTR PFN_vkVoidFunction VKAPI_CALL
       const char* pName);
 ```
 
-This function has very similar semantics to `vkGetInstanceProcAddr`.
-`vk_icdGetInstanceProcAddr` returns valid function pointers for all the
-global-level and instance-level Vulkan functions, and also for
-`vkGetDeviceProcAddr`.
+This function has the same semantics as `vkGetInstanceProcAddr`.
+
```

Comment:
I think a note should be placed here stating that this is the fucntion that the loader uses to perform the same functionality as vkGetInstanceProcAddr but on a driver basis.

Reply:
I've yet to test the actual code, but I have some concern with the new definition. It requires the ICD to support device-level functions via `vk_icdGetInstanceProcAddr`. This may require the ICD to do its own trampolining (e.g. to efficiently support differing physical device models that are present on the system) rather than piggy-back on the loader's ability to do so.

Reply:
From my understanding, the only time a driver will have device function queried with GIPA will be when it is an unknown function to the loader. Anything known to the loader when queried through GIPA will return the appropriate loader trampoline, and if no trampoline is needed, the first layer in the chain.

Also, this PR doesn't change the loader behavior except to not load 2 functions with GIPA that are subsequently loaded by GDPA (during device creation). I do not know of any impacts to existing behavior. I don't know how those 2 functions that will now be null could be reached since they are loaded into an instance level dispatch table that is only used for instance & physical device level functions.

My stance is that drivers are written against the Vulkan spec, and will generally follow the spec first and foremost. I would expect that drivers handle device querying from GIPA simply cause the spec requires it, and many drivers are tested sans the loader initially. Looking at mesa, the common code for mesa follows the spec in this regard. (of course, that is not all drivers, just several of them).

Reply:
You certainly bring up a good point wrt unknown functions. That case hadn't crossed my mind.

Prior to physical-device level functions being introduced I'd have a case to argue that the loader shouldn't expose anything instance-level it doesn't support, thus an unknown function had to be a device function. Today, it wouldn't affect this discussion regardless.

Reply:
Sorry if this is just a misunderstanding, but there have always been physical device level functions. Do you mean 'prior to allowing unknown physical device functions'? That is what the reason vkGetPhysicalDeviceProcAddr was added, not that there weren't functions which took a physical device as its dispatchable handle.

I have a feeling that the documentation change should be something I run by the System Integration WG before merging, as a procedural "in case my assumptions are flat out wrong". While I would like to merge this PR as is, I'm open to splitting it up into 2, one for the code changes & the other to handle the documentation change. Let me know if you feel that is needed.

Reply:
> Do you mean 'prior to allowing unknown physical device functions'?
That's actually exactly what I intended to say. Thanks!

I believe I've spoken ahead of myself earlier. My apologies.

Take this scenario as an example:
* App requests via `vkGetInstanceProcAddr` a command unknown to the loader.
*  The loader doesn't expose instance-level functionality it does not support, therefore not an instance-level function. (This is an assumption I'm making. I don't know if it's currently true, correct me if I'm wrong, please.)
* The ICD's `vk_icdGetPhysicalDeviceProcAddr` returns NULL, therefore not a physical-device level function.
* Thus the only possibility remains that the unknown command is a device-level function.

Could we therefore not rely on the ICD's `vkGetDeviceProcAddr` for unknown commands we know to be device-level, and avoid the ICD's need to implement trampolines for device-level commands queried via `vk_icdGetInstanceProcAddr`?

To clarify, I'm referring here to what the loader is allowed to do and assume under the Loader-ICD interface contract, not necessarily what the loader actually does today.

The code changes themselves look good at a first glance. I'll take the opportunity to test them tomorrow evening. Thanks!

Reply:
We can't use `vkGetDeviceProcAddr` in the case of an unknown device function being queried with `vkGetInstanceProcAddr`, because there is now way to know *which* device is being queried against.

The 'device' handle is known when the application calls the function, because the function the loader returned is a trampoline that unwraps the VkDevice that is passed in to figure out which driver's device function to call. The loader keeps a separate list of unknown functions for each driver, which is how the loader is able to call the right driver when the function is called.

Reply:
> [...] there is now way to know which device is being queried against.

aha!

> The loader keeps a separate list of unknown functions for each driver, which is how the loader is able to call the right driver when the function is called.

Right, so the loader must create a trampoline. To continue the discussion, could this list not be made per-device rather than per-driver, therefore allow the use of the ICD's `vkGetDeviceProcAddr` for the pointer that the loader's trampoline branches to?

Nevertheless, I think that such a change might be out-of-scope for resolving this issue. Nevertheless, your proposed documentation change is clear and doesn't leave room for misinterpretation, which is always a boon. 👍

Reply:
To answer you question, I misspoke about the loader having per-driver extension lists. I just looked at the code and it turns out it does what you want already. I saw a `for loop over the ICD list and thought "must do it per driver", not noticing the `while` loop inside said `for` loop. On top of that, while the loader uses `vkGetInstanceProcAddr` to determine if any drivers support an unknown function, it actually uses `vkGetDeviceProcAddr` to get the actual function pointer to use.

I'll address Mark's comment above at some future time then consider it approved and will merge it then.

Side note:
This whole mess could have been avoided if `vkGetInstanceProcAddr` didn't support device function querying. But I don't get to change history, so workarounds and very confusing implementation details is what we live with.

Reply:
All right. I'll play around with this tomorrow, especially wrt unknown functions, and get back to you with my findings.

Thank you for your time looking into this! It is much appreciated.

Reply:
Mark, I fixed it.
Polarina, since you haven't gotten back, I'm going to go ahead and merge it, since the code changes were approved. Feel free to comment here or open a follow up issue if things aren't correct.